src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,972 @@
+ /**
+ * @file neat.hpp
+ * @author Bang Liu
+ *
+ * Definition of NEAT class.
+ */
+#ifndef MLPACK_METHODS_NE_NEAT_HPP
+#define MLPACK_METHODS_NE_NEAT_HPP
+
+#include <cstddef>
+#include <cstdio>
+#include <numeric>
+
+#include <mlpack/core.hpp>
+
+#include "link_gene.hpp"
+#include "neuron_gene.hpp"
+#include "genome.hpp"
+#include "species.hpp"
+#include "population.hpp"
+#include "tasks.hpp"
+#include "parameters.hpp"
+
+namespace mlpack {
+namespace ne {
+
+struct LinkInnovation {
+  ssize_t fromNeuronId;
+  ssize_t toNeuronId;
+  ssize_t newLinkInnovId;
+};
+
+struct NeuronInnovation {
+  ssize_t splitLinkInnovId;
+  ssize_t newNeuronId;
+  ssize_t newInputLinkInnovId;
+  ssize_t newOutputLinkInnovId;
+};
+
+/**
+ * This class implements  NEAT algorithm.
+ */
+template<typename TaskType>
+class NEAT {
+ public:
+  // Parametric constructor.
+  NEAT(TaskType task, Genome& seedGenome, Parameters& params) {
+    aTask = task;
+    aSeedGenome = seedGenome;
+    aPopulationSize = params.aPopulationSize;
+    aMaxGeneration = params.aMaxGeneration;
+    aCoeffDisjoint = params.aCoeffDisjoint;
+    aCoeffWeightDiff = params.aCoeffWeightDiff;
+    aCompatThreshold = params.aCompatThreshold;
+    aStaleAgeThreshold = params.aStaleAgeThreshold;
+    aCrossoverRate = params.aCrossoverRate;
+    aCullSpeciesPercentage = params.aCullSpeciesPercentage;
+    aMutateWeightProb = params.aMutateWeightProb;
+    aPerturbWeightProb = params.aPerturbWeightProb;
+    aMutateWeightSize = params.aMutateWeightSize;
+    aMutateAddLinkProb = params.aMutateAddLinkProb;
+    aMutateAddRecurrentLinkProb = params.aMutateAddRecurrentLinkProb;
+    aMutateAddLoopLinkProb = params.aMutateAddLoopLinkProb;
+    aMutateAddNeuronProb = params.aMutateAddNeuronProb;
+    aMutateEnabledProb = params.aMutateEnabledProb;
+    aMutateDisabledProb = params.aMutateDisabledProb;
+  }
+
+  // Destructor.
+  ~NEAT() {}
+
+  // Check whether a link innovation already exist.
+  ssize_t CheckLinkInnovation(ssize_t fromNeuronId, ssize_t toNeuronId) {
+    for (ssize_t i=0; i<aLinkInnovations.size(); ++i) {
+      if (aLinkInnovations[i].fromNeuronId == fromNeuronId && 
+          aLinkInnovations[i].toNeuronId == toNeuronId) {
+        return i;
+      }
+    }
+    
+    return -1;  // -1 means no match found, a new innovation.
+  }
+
+  // Check whether a neuron innovation already exist.
+  ssize_t CheckNeuronInnovation(ssize_t splitLinkInnovId) {
+    for (ssize_t i=0; i<aNeuronInnovations.size(); ++i) {
+      if (aNeuronInnovations[i].splitLinkInnovId == splitLinkInnovId) {
+        return i;
+      }
+    }
+
+    return -1;
+  }
+
+  // Add a new link innovation.
+  LinkInnovation AddLinkInnovation(ssize_t fromNeuronId, ssize_t toNeuronId) {
```

Comment:
I think it's a good idea to use a reference here:

``
 AddLinkInnovation(ssize_t fromNeuronId, ssize_t toNeuronId, LinkInnovation& linkInnov)
``